README.md
# optrita 
Optimization algorithms by Rita ♥️

*pip3 install* with `! pip3 install -i https://test.pypi.org/simple/ optrita==0.0.3` 
<hr>

## Table of Contents
- [**Non-linear Optimization**](#non-linear-optimization)
    - [Parameters](#parameters)
    - [Linesearch](#linesearch-bls)
    - [Unconstrained optimization](#uncontrained-optimization)
        - [First derivative methods](#first-derivative-methods)
            - [Gradient Method](#gradient-method-gm)
            - [Conjugate Gradient Method](#conjugate-gradient-method-cgm)
            - [Broyden-Fletcher-Goldfarb-Shanno (BFGS) Method](#broyden-fletcher-goldfarb-shanno-bfgs-method-bfgs)
        - [Second derivative methods](#second-deriavtive-methods)
            - [Newton Method](#newton-method-newton)
            - [Modified Newton](#modified-newton)
                - [Modified Newton's method based on Spectral Decomposition](#modified-newtons-method-based-on-spectral-decomposition-mnsd)
                - [Modified Newton's method based on the Cholesky factorization](#modified-newtons-method-based-on-the-cholesky-factorization-mncf)
<hr>

## Non-linear Optimization

### Parameters

### Linesearch `BLS`

### Unconstrained Optimization
<hr>

#### ~ First derivative methods ~
<hr>

### Gradient Method `GM`
#### Optimize with `GM`
```python
import numpy as np 
import optrita as opt

# Define the function to minimize and its gradient:
def f(x): return -x[0]**3 - (1/2)*x[0]**2 + (1/2)*x[1]**2 + (1/4)*x[0]*x[1] + x[0]
def g(x): return np.array([-3*x[0]**2 - x[0] + (1/4)*x[1] + 1,
                           x[1] + (1/4)*x[0]])

# Define the initial point:
x = np.array([[0.4, 0.6]]).T

# Define the linesearch parameters:
BLS_params = dict(alpha_max = 2.0,
                  alpha_min = 1e-3,
                  rho = 0.5,
                  c1 = 0.01, c2 = 0.45,
                  strong_wolfe = True, out = 0)

# Optimize!
Xk_GM, info_GM = opt.GM(x, f, g, BLS_params, eps = 1e-6, kmax = 1500, precision = 6)

```
<hr>

### Conjugate Gradient Method `CGM`
#### Optimize with `CGM`
```python
import numpy as np 
import optrita as opt

# Define the function to minimize and its gradient:
def f(x): return -x[0]**3 - (1/2)*x[0]**2 + (1/2)*x[1]**2 + (1/4)*x[0]*x[1] + x[0]
def g(x): return np.array([-3*x[0]**2 - x[0] + (1/4)*x[1] + 1,
                           x[1] + (1/4)*x[0]])

# Define the initial point:
x = np.array([[0.4, 0.6]]).T

# Define the linesearch parameters:
BLS_params = dict(alpha_max = 2.0,
                  alpha_min = 1e-3,
                  rho = 0.5,
                  c1 = 0.01, c2 = 0.45,
                  strong_wolfe = True, out = 0)

# Optimize!
Xk_CGM, info_CGM = opt.CGM(x, f, g, BLS_params, iCG = 1, iRC = 1, eps = 1e-6, kmax = 1500, precision = 6,nu = 0.1)

```
<hr>

### Broyden-Fletcher-Goldfarb-Shanno (BFGS) Method `BFGS`
#### Optimize with `BFGS`
```python
import numpy as np 
import optrita as opt

# Define the function to minimize and its gradient:
def f(x): return -x[0]**3 - (1/2)*x[0]**2 + (1/2)*x[1]**2 + (1/4)*x[0]*x[1] + x[0]
def g(x): return np.array([-3*x[0]**2 - x[0] + (1/4)*x[1] + 1,
                           x[1] + (1/4)*x[0]])

# Define the initial point:
x = np.array([[0.4, 0.6]]).T

# Define the linesearch parameters:
BLS_params = dict(alpha_max = 2.0,
                  alpha_min = 1e-3,
                  rho = 0.5,
                  c1 = 0.01, c2 = 0.45,
                  strong_wolfe = True, out = 0)

# Optimize!
Xk_BFGS, info_BFGS = opt.BFGS(x, f, g, BLS_params, eps = 1e-6, kmax = 1500, precision = 6)

```
<hr>

#### ~ Second derivative methods ~
<hr>

### Newton Method `newton`
The rationale behind the Newton’s method is to subtitute, at each iterate <a href="https://www.codecogs.com/eqnedit.php?latex=x^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^k" title="x^k" /></a>, the true function <a href="https://www.codecogs.com/eqnedit.php?latex=f" target="_blank"><img src="https://latex.codecogs.com/gif.latex?f" title="f" /></a> by the quadratic approximation <a href="https://www.codecogs.com/eqnedit.php?latex=f^k_Q" target="_blank"><img src="https://latex.codecogs.com/gif.latex?f^k_Q" title="f^k_Q" /></a> around <a href="https://www.codecogs.com/eqnedit.php?latex=x_k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x_k" title="x_k" /></a>:

<a href="https://www.codecogs.com/eqnedit.php?latex=f(x^k&plus;p)&space;\approx&space;f^k_Q(p)&space;=&space;f^k&space;&plus;&space;\nabla&space;f^{k^T}\cdot&space;p&space;&plus;&space;\frac{1}{2}p^T&space;\cdot&space;\nabla^2&space;f^k&space;\cdot&space;p" target="_blank"><img src="https://latex.codecogs.com/gif.latex?f(x^k&plus;p)&space;\approx&space;f^k_Q(p)&space;=&space;f^k&space;&plus;&space;\nabla&space;f^{k^T}\cdot&space;p&space;&plus;&space;\frac{1}{2}p^T&space;\cdot&space;\nabla^2&space;f^k&space;\cdot&space;p" title="f(x^k+p) \approx f^k_Q(p) = f^k + \nabla f^{k^T}\cdot p + \frac{1}{2}p^T \cdot \nabla^2 f^k \cdot p" /></a>

and then, to define the next iterate <a href="https://www.codecogs.com/eqnedit.php?latex=x^{k&plus;1}$&space;as&space;$x^{k&plus;1}&space;=&space;x^k&space;&plus;p^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^{k&plus;1}$&space;as&space;$x^{k&plus;1}&space;=&space;x^k&space;&plus;p^k" title="x^{k+1}$ as $x^{k+1} = x^k +p^k" /></a> with <a href="https://www.codecogs.com/eqnedit.php?latex=p^k_N&space;:=&space;\arg&space;\min&space;{f^k_Q(p)}&space;=&space;-&space;\nabla^2f^{k^{-1}}&space;\cdot&space;\nabla&space;f^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?p^k_N&space;:=&space;\arg&space;\min&space;{f^k_Q(p)}&space;=&space;-&space;\nabla^2f^{k^{-1}}&space;\cdot&space;\nabla&space;f^k" title="p^k_N := \arg \min {f^k_Q(p)} = - \nabla^2f^{k^{-1}} \cdot \nabla f^k" /></a>

<a href="https://www.codecogs.com/eqnedit.php?latex=x^{k&plus;1}&space;\leftarrow&space;x^k&space;-&space;\nabla^2f^{k^{-1}}&space;\cdot&space;\nabla&space;f^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^{k&plus;1}&space;\leftarrow&space;x^k&space;-&space;\nabla^2f^{k^{-1}}&space;\cdot&space;\nabla&space;f^k" title="x^{k+1} \leftarrow x^k - \nabla^2f^{k^{-1}} \cdot \nabla f^k" /></a>

#### ~ Global Convergence ~

No guarantees! If <a href="https://www.codecogs.com/eqnedit.php?latex=x^0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^0" title="x^0" /></a> is far from a strict minimizer <a href="https://www.codecogs.com/eqnedit.php?latex=x^*$,&space;$p^k_N" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^*$,&space;$p^k_N" title="x^*$, $p^k_N" /></a> may not be a descent direction.

#### ~ Local Convergence ~
if the starting point <a href="https://www.codecogs.com/eqnedit.php?latex=x^0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^0" title="x^0" /></a> is sufficiently close to <a href="https://www.codecogs.com/eqnedit.php?latex=x^*" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^*" title="x^*" /></a> then:
+ <a href="https://www.codecogs.com/eqnedit.php?latex=\{x^k\}&space;\rightarrow&space;x^*" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\{x^k\}&space;\rightarrow&space;x^*" title="\{x^k\} \rightarrow x^*" /></a>
+ The order of convergence of <a href="https://www.codecogs.com/eqnedit.php?latex=\{x^k\}^\infty_{k=0}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\{x^k\}^\infty_{k=0}" title="\{x^k\}^\infty_{k=0}" /></a> is quadratic.
+ The sequence of gradient norms <a href="https://www.codecogs.com/eqnedit.php?latex=\{\|&space;\nabla&space;f^k&space;\|\}^\infty_{k=0}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\{\|&space;\nabla&space;f^k&space;\|\}^\infty_{k=0}" title="\{\| \nabla f^k \|\}^\infty_{k=0}" /></a> converges quadratically to zero.

#### Optimize with `newton`
```python
import numpy as np 
import optrita as opt

# Define the function to minimize, its gradient and its hessian:
def f(x): return -x[0]**3 - (1/2)*x[0]**2 + (1/2)*x[1]**2 + (1/4)*x[0]*x[1] + x[0]
def g(x): return np.array([-3*x[0]**2 - x[0] + (1/4)*x[1] + 1,
                           x[1] + (1/4)*x[0]])
def h(x): return np.array([[-6*x[0]-1, 1/4],
                           [1/4, 1]])

# Define the initial point:
x = np.array([[0.1, 0.6]]).T

# Define the linesearch parameters:
BLS_params = dict(alpha_max = 1.0,
                 alpha_min = 1e-3,
                 rho = 0.5,
                 c1 = 0.01, c2 = 0.45,
                 strong_wolfe = False, out = 0)

# Optimize!
Xk_N, info_N = opt.newton(x, f, g, h, BLS_params, eps = 10e-5, kmax = 100, precision = 6)

```
<hr>
    
### Modified Newton
Modified Newton’s method where a positive definite approximation to the true hessian <a href="https://www.codecogs.com/eqnedit.php?latex=\nabla^2f^{k}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\nabla^2f^{k}" title="\nabla^2f^{k}" /></a> is used at each iteration, so as to make <a href="https://www.codecogs.com/eqnedit.php?latex=d^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?d^k" title="d^k" /></a> a descent direction:

<a href="https://www.codecogs.com/eqnedit.php?latex=B^k&space;=&space;\nabla^2f^{k}&space;&plus;&space;E^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?B^k&space;=&space;\nabla^2f^{k}&space;&plus;&space;E^k" title="B^k = \nabla^2f^{k} + E^k" /></a>

Near the optimizer <a href="https://www.codecogs.com/eqnedit.php?latex=x^*" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^*" title="x^*" /></a> we want the <a href="https://www.codecogs.com/eqnedit.php?latex=E^k=0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?E^k=0" title="E^k=0" /></a> and the step length <a href="https://www.codecogs.com/eqnedit.php?latex=\alpha=1" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\alpha=1" title="\alpha=1" /></a> to exhibit Newton's quadratic convergence.

#### ~ Global convergence ~
If we start at <a href="https://www.codecogs.com/eqnedit.php?latex=x^0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^0" title="x^0" /></a> and the bounded modified condition holds:

<a href="https://www.codecogs.com/eqnedit.php?latex=\kappa(B^k)&space;\leq&space;C,&space;\space&space;C>0,&space;\forall&space;k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\kappa(B^k)&space;\leq&space;C,&space;\space&space;C>0,&space;\forall&space;k" title="\kappa(B^k) \leq C, \space C>0, \forall k" /></a>

the algorithm converges to a stationary point.

#### ~ Local convergence ~
Given that
+ The SOSC are satisfied at <a href="https://www.codecogs.com/eqnedit.php?latex=x^*" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^*" title="x^*" /></a>.
+ The Hessian <a href="https://www.codecogs.com/eqnedit.php?latex=\nabla^2&space;f" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\nabla^2&space;f" title="\nabla^2 f" /></a> is Lipschitz continuous in a neighborhood of <a href="https://www.codecogs.com/eqnedit.php?latex=x^*" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x^*" title="x^*" /></a>.
+ <a href="https://www.codecogs.com/eqnedit.php?latex=E^k&space;=&space;0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?E^k&space;=&space;0" title="E^k = 0" /></a> for <a href="https://www.codecogs.com/eqnedit.php?latex=k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?k" title="k" /></a> large enough.

Then the order of convergence is quadratic. 

If <a href="https://www.codecogs.com/eqnedit.php?latex=\nabla^2&space;f^*" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\nabla^2&space;f^*" title="\nabla^2 f^*" /></a> is close to singular, the first point may not hold and then the convergence might be only linear.
<hr>

### Modified Newton's method based on Spectral Decomposition `MNSD`
The following function positifies any Hessian by decomposing it in the spectral decomposition and applying:

<a href="https://www.codecogs.com/eqnedit.php?latex=B^k_{MN-SD}&space;=&space;Q&space;\cdot&space;diag&space;(\max(\delta,&space;\lambda_i))&space;\cdot&space;Q^T" target="_blank"><img src="https://latex.codecogs.com/gif.latex?B^k_{MN-SD}&space;=&space;Q&space;\cdot&space;diag&space;(\max(\delta,&space;\lambda_i))&space;\cdot&space;Q^T" title="B^k_{MN-SD} = Q \cdot diag (\max(\delta, \lambda_i)) \cdot Q^T" /></a>

It assures that the direction <a href="https://www.codecogs.com/eqnedit.php?latex=d^k_{MN-SD}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?d^k_{MN-SD}" title="d^k_{MN-SD}" /></a> is a descent direction.
<hr>

### Modified Newton's method based on the Cholesky factorization `MNCF`
Cholesky factorization <a href="https://www.codecogs.com/eqnedit.php?latex=A&space;=&space;R^T&space;\cdot&space;R" target="_blank"><img src="https://latex.codecogs.com/gif.latex?A&space;=&space;R^T&space;\cdot&space;R" title="A = R^T \cdot R" /></a> exists if <a href="https://www.codecogs.com/eqnedit.php?latex=A" target="_blank"><img src="https://latex.codecogs.com/gif.latex?A" title="A" /></a> is a symmetric and definite positive. The following method is based on an iterative modification <a href="https://www.codecogs.com/eqnedit.php?latex=\nabla^2&space;f^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\nabla^2&space;f^k" title="\nabla^2 f^k" /></a> that guarantees the positive definiteness of <a href="https://www.codecogs.com/eqnedit.php?latex=B^k" target="_blank"><img src="https://latex.codecogs.com/gif.latex?B^k" title="B^k" /></a>:

<a href="https://www.codecogs.com/eqnedit.php?latex=B^k&space;=&space;\nabla^2&space;f^k&space;&plus;&space;\tau&space;\cdot&space;I" target="_blank"><img src="https://latex.codecogs.com/gif.latex?B^k&space;=&space;\nabla^2&space;f^k&space;&plus;&space;\tau&space;\cdot&space;I" title="B^k = \nabla^2 f^k + \tau \cdot I" /></a>

With strictly increasingly <a href="https://www.codecogs.com/eqnedit.php?latex=\tau&space;\geq&space;0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\tau&space;\geq&space;0" title="\tau \geq 0" /></a> is going to be tried until the factorization succeeds for some <a href="https://www.codecogs.com/eqnedit.php?latex=\tau" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\tau" title="\tau" /></a>.
<hr>

## License

MIT © `3omni`


CHANGES.txt
v0.0.3, oct 2019 -- Third release


optrita/linesearch/.ipynb_checkpoints/base-checkpoint.py
import numpy as np
import pandas as pd
import sympy as sy
import pandas as pd
import matplotlib.pyplot as plt

#------------------------------------------------------------------------------
# BLS algorithm 
#------------------------------------------------------------------------------
def BLS(x, d, f, g, alpha_max = 1.0, alpha_min = 1e-6, 
        rho = 0.5, c1 = 0.1, c2 = 0.5, 
        strong_wolfe = False, out = True):
    """
    Parameters
    ----------
    x : initial point
    d : initial direction
    f : function
    g : gradient function
    alpha_max : initial value of alpha
    alpha_min : final minimum value of alpha
    rho, c1, c2 parameters
    strong_wolfe : indicator for Wolfe conditions
        - True : ask for Strong Wolfe conditions
        - False : (default) just Wolfe conditions
    out : print out logs. By default it is set to true
    
    Returns
    -------
    alph : the optimal alpha
    W_out : the satisfied Wolfe conditions
    
    """
    def WC1(al):
        return f(x+alph*d) <= f(x) + c1*g(x).T @ d*alph
    def WC2(al):
        return g(x+alph*d).T @ d >= c2*g(x).T @ d
    def SWC2(al):
        return np.abs(g(x+alph*d).T@d) <= c2*np.abs(g(x).T@d)
    # =========================================================================== #
    al, rh = sy.symbols('α ρ');
    if out == True:
        print("# ==================================================================== #")
        print("Initial point:", x.T)
        print("Initial direction:", d.T)
        print("Backtracking Linear Search Hyperparameters:")
        print("  ", al, " max:", alpha_max)
        print("  ", al, " min:", alpha_min)
        print("  ", rh, ":", rho)
        print("   c1:", c1)
        print("   c2:", c2)
        print("Asked for Strong Wolfe conditions: ", strong_wolfe)
        print("# ==================================================================== #")
    # =========================================================================== #
    # Acceptability conditions:
    def W(alph):
        if strong_wolfe:
            return WC1(alph) and SWC2(alph)
        else:
            return WC1(alph) and WC2(alph)
        
    # Initialization algorithm:
    alph = alpha_max
    
    # Loop: while the output condition is false ~.
    it_max = 100
    if out: print("Starting search...")
    it = 0
    while not W(alph) and alph > alpha_min:
        if out == True: print(al,"(",it, ")" " = ", alph, sep="")
        alph = rho*alph
        it += 1
        if it > it_max:
            break
    if out == True: 
        print(al,"(",it, ")" " = ", alph, sep="")
        print("# ==================================================================== #")
        if alph <= alpha_min:
            print("Stopped iterating by ",al, " min. NOT FOUND.")
        print("Output:", al,"=",alph)
    
    W_out = 0
    if WC1(alph):
        W_out = 1
    if WC1(alph) and WC2(alph):
        W_out = 2
    if WC1(alph) and SWC2(alph):
        W_out = 3
    if out == True:   
        # print result WC
        if W_out == 0:
            print("Does not satisfy any Wolfe Condition.")
        elif W_out == 1:
            print("Satisfies Wolfe Condition 1: Sufficient decrease.")
        elif W_out == 2:
            print("""Satisfies Wolfe Condition 1 and Wolfe Condition 2:\nSufficient decrease and Curvature condition.
            """)
        elif W_out == 3:
            print("""Satisfies Wolfe Condition 1 and Strong Wolfe Condition 2:\nSufficient decrease and (Strong) Curvature condition.
            """)
    return alph, W_out

# FIRST DERIVATIVE METHODS
#------------------------------------------------------------------------------
# Gradient method for any function f
#------------------------------------------------------------------------------
def GM(x, f, g, BLS_params, eps = 1e-6, kmax = 1500, precision = 6):
    """
    Parameters
    ----------
    x : initial point
    f : function
    g : gradient function
    BLS_params : dictionary with BLS parameters
    eps : epsilon for stopping condition on gradient norm
    kmax : maximum number of iterations
    precision : decimal precision on gradient norm
    
    Returns
    -------
    Xk : x at each iteration
    data : additional information on each iteration (alpha, gradient norm)
    
    """
    gradient_norm = round(np.linalg.norm(g(x)), precision)
    info = [[np.NaN, gradient_norm]]
    Xk = []
    dimension = len(x)
    for i in range(0, dimension):
        Xk.append([x[i].squeeze()])
    
    k = 0
    print("[GM] Initial:", x.T)
    # ================================================================================================ #  
    while np.linalg.norm(g(x)) > eps and k < kmax:
        d = -g(x)
        alpha, _ = BLS(x, d, f, g, **BLS_params)   # Backtracking Line Search
        x = x + alpha*d
        k += 1
        gradient_norm = np.round(np.linalg.norm(g(x)), precision)
        info.append([alpha, gradient_norm])
        for i in range(0, dimension):
            Xk[i].append(x[i].squeeze())
       # ============================================================================================== #  
    # OUTPUT:
    print("[GM] Final:",x.T)
    print("[GM] Iterations:", k)
    data = pd.DataFrame(info, columns=["alpha", "||g(x)||"], dtype=np.float)
    return Xk, data

#------------------------------------------------------------------------------
# Conjugate Gradient method
#------------------------------------------------------------------------------
def CGM(x, f, g, BLS_params, iCG, iRC, eps = 1e-6, kmax = 1500, precision = 6, nu = None):
    """
    Parameters
    ----------
    x : initial point
    f : function
    g : gradient function
    BLS_params : dictionary with BLS parameters
    eps : epsilon for stopping condition on gradient norm
    kmax : maximum number of iterations
    precision : decimal precision on gradient norm
    iCG : type of CGM. iCG 1 is FR, 2 is PR
    iRC : type of restart condition
    nu : needed for a restart condition
    
    Returns
    -------
    Xk : x at each iteration
    data : additional information on each iteration (alpha, gradient norm)
    
    """
    gradient_norm = round(np.linalg.norm(g(x)), precision)
    info = [[np.NaN, gradient_norm]]
    Xk = []
    dimension = len(x)
    for i in range(0, dimension):
        Xk.append([x[i].squeeze()])
    d = -g(x)
    k = 0
    print("[CGM] Initial:", x.T)
    if BLS_params['strong_wolfe'] != True:
        print("WARNING: ⚠️ in order to guarantee a descent direction, the step length α must satisfy SWC.")
    if BLS_params['c2'] >= 0.5:
        print("WARNING: ⚠️ in order to guarantee a descent direction, c2 must be smaller than 0.5.")
    # ======================================================================================================= #  
    while np.linalg.norm(g(x)) > eps and k < kmax:
        alpha, _ = BLS(x, d, f, g, **BLS_params)
        x, x_prev = x + alpha*d, x
        # ======================================================================================================= #  
        # CGM variants
        if iCG == 1:
            beta = (g(x).T @ g(x)) / (g(x_prev).T @ g(x_prev))
        elif iCG == 2:
            beta = max(0, g(x).T @ (g(x) - g(x_prev)) / (g(x_prev).T @ g(x_prev)))
        else:
            raise TypeError("iCG should be 1 (Fletcher-Reeves) or 2 (Polak-Ribière)")
        # ======================================================================================================= # 
        # Restart conditions
        if iRC > 0 and nu is None:
            raise TypeError(f"nu is a necessary parameter with iRC equal to {iRC}")
        if (iRC == 1 and k % nu == 0 or
                iRC == 2 and g(x).T @ g(x_prev) / np.linalg.norm(g(x))**2 > nu or
                k == 0):
            d = -g(x)
        else:
            d = -g(x) + beta*d
        # ======================================================================================================= # 
        k += 1
        gradient_norm = round(np.linalg.norm(g(x)), precision)
        info.append([alpha, gradient_norm])
        for i in range(0, dimension):
            Xk[i].append(x[i].squeeze())
    # ======================================================================================================= #
    # OUTPUT:
    print("[CGM] Final:",x.T)
    print("[CGM] Iterations:", k)
    data = pd.DataFrame(info, columns=["alpha", "||g(x)||"], dtype=np.float)
    return Xk, data


#------------------------------------------------------------------------------
# BFGS method
#------------------------------------------------------------------------------
def BFGS(x, f, g, BLS_params, eps = 1e-6, kmax = 1500, precision = 6):
    """
    Parameters
    ----------
    x : initial point
    f : function
    g : gradient function
    BLS_params : dictionary with BLS parameters
    eps : epsilon for stopping condition on gradient norm
    kmax : maximum number of iterations
    precision : decimal precision on gradient norm
    
    Returns
    -------
    Xk : x at each iteration
    data : additional information on each iteration (alpha, gradient norm)
    
    """
    gradient_norm = round(np.linalg.norm(g(x)), precision)
    info = [[np.NaN, gradient_norm]]
    Xk = []
    dimension = len(x)
    for i in range(0, dimension):
        Xk.append([x[i].squeeze()])
    
    H = I = np.identity(len(g(x)))
    k = 0
    print("[BFGS] Initial:", x.T)
    while np.linalg.norm(g(x)) > eps and k < kmax:
        d = -H @ g(x)
        alpha, _ = BLS(x, d, f, g, **BLS_params)
        x, x_prev = x + alpha*d, x
        s = x - x_prev
        y = g(x) - g(x_prev)
        rho = 1 / ((y).T @ s)
        H = (I - rho * s @ y.T) @ H @ (I - rho * y @ (s.T)) + rho * s @ s.T
        k += 1
        gradient_norm = round(np.linalg.norm(g(x)), precision)
        info.append([alpha, gradient_norm])
        for i in range(0, dimension):
            Xk[i].append(x[i].squeeze())
        
    # ================================================================================================ #
    # OUTPUT:
    print("[BFGS] Final:",x.T)
    print("[BFGS] Iterations:", k)
    data = pd.DataFrame(info, columns=["alpha", "||g(x)||"], dtype=np.float)
    return Xk, data

# SECOND DERIVATIVE METHODS
#------------------------------------------------------------------------------
# Newton method
#------------------------------------------------------------------------------
def newton(x, f, g, h, BLS_params, eps = 1e-6, kmax = 1500, precision = 6):
    """
    Parameters
    ----------
    x : initial point
    f : function
    g : gradient function
    h : hessian
    BLS_params : dictionary with BLS parameters
    eps : epsilon for stopping condition on gradient norm
    kmax : maximum number of iterations
    precision : decimal precision on gradient norm
    
    Returns
    -------
    Xk : x at each iteration
    data : additional information on each iteration (alpha, gradient norm)
    
    """
    gradient_norm = np.round(np.linalg.norm(g(x)), precision)
    info = [[np.NaN, gradient_norm]]
    Xk = []
    dimension = len(x)
    for i in range(0, dimension):
        Xk.append([x[i].squeeze()])
    
    print("[N] Initial:", x.T)
    k = 0
    while np.linalg.norm(g(x)) > eps and k < kmax:
        lamb, bibi = np.linalg.eig(h(x.squeeze()))
        print(lamb, "\t")
        d = -np.linalg.inv(h(x.squeeze())) @ g(x)
        x = x + d
        k += 1
        gradient_norm = round(np.linalg.norm(g(x)), precision)
        info.append([alpha, gradient_norm])
        for i in range(0, dimension):
            Xk[i].append(x[i].squeeze())
    # ================================================================================================ #
    # OUTPUT:
    print("[N] Final:",x.T)
    print("[N] Iterations:", k)
    data = pd.DataFrame(info, columns=["alpha", "||g(x)||"], dtype=np.float)
    return Xk, data

#------------------------------------------------------------------------------
# Modified Newton's method based on Spectral Decomposition
#------------------------------------------------------------------------------
def positify_hessian(H, delta = 0.01):
    """
    Paramters
    ---------
    H : matrix
    delta : clip factor
    
    Returns
    -------
    Positified matrix H
    
    """
    lambdas, Q = np.linalg.eig(H)
    lambdas = lambdas.clip(delta)
    return Q @ np.diag(lambdas) @ Q.T

def MNSD(x, f, g, h, BLS_params, eps = 1e-6, kmax = 1500, precision = 6):
    """
    Parameters
    ----------
    x : initial point
    f : function
    g : gradient function
    h : hessian
    BLS_params : dictionary with BLS parameters
    eps : epsilon for stopping condition on gradient norm
    kmax : maximum number of iterations
    precision : decimal precision on gradient norm
    
    Returns
    -------
    Xk : x at each iteration
    data : additional information on each iteration (alpha, gradient norm)
    
    """
    gradient_norm = np.round(np.linalg.norm(g(x)), precision)
    info = [[np.NaN, gradient_norm]]
    Xk = []
    dimension = len(x)
    for i in range(0, dimension):
        Xk.append([x[i].squeeze()])
    
    print("[MN-SD] Initial:", x.T)
    k = 0
    while np.linalg.norm(g(x)) > eps and k < kmax:
        print(h(x.squeeze()), "\t")
        B = positify_hessian(h(x.squeeze()))
        print(B, "\t")
        d = -np.linalg.inv(B) @ g(x)
        alpha, _ = BLS(x, d, f, g, **BLS_params)
        x = x + alpha*d
        k += 1
        
        gradient_norm = np.round(np.linalg.norm(g(x)), precision)
        info.append([alpha, gradient_norm])
        for i in range(0, dimension):
            Xk[i].append(x[i].squeeze())
    # ================================================================================================ #
    # OUTPUT:
    print("[MN-SD] Final:",x.T)
    print("[MN-SD] Iterations:", k)
    data = pd.DataFrame(info, columns=["alpha", "||g(x)||"], dtype=np.float)
    return Xk, data

#------------------------------------------------------------------------------
# Modified Newton's method based on the Cholesky factorization
#------------------------------------------------------------------------------
def CMI(H, t = 0, k = 0): # recursive
    """
    Parameters
    ----------
    H : hessian
    t : factor
    k : iteration
    
    Returns
    -------
    Factorized matrix H with t factor at iteration k

    """
    try:
        R = np.linalg.cholesky(H)
        return R @ R.T
    except np.linalg.LinAlgError:
        # print("⚠️ Matrix is not positive definite at it", k)
        l_UB = np.linalg.norm(H, ord='fro')
        k = k + 1
        t = (1.01 - 1/2**k)*l_UB
        return CMI(H + t*np.eye(len(H)), t, k) 
    
def MNCF(x, f, g, h, BLS_params, eps = 1e-6, kmax = 1500, precision = 6):
    """
    Parameters
    ----------
    x : initial point
    f : function
    g : gradient function
    h : hessian
    BLS_params : dictionary with BLS parameters
    eps : epsilon for stopping condition on gradient norm
    kmax : maximum number of iterations
    precision : decimal precision on gradient norm
    
    Returns
    -------
    Xk : x at each iteration
    data : additional information on each iteration (alpha, gradient norm)
    
    """
    gradient_norm = np.round(np.linalg.norm(g(x)), precision)
    info = [[np.NaN, gradient_norm]]
    Xk = []
    dimension = len(x)
    for i in range(0, dimension):
        Xk.append([x[i].squeeze()])
    
    print("[MN-CF] Initial:", x.T)
    k = 0
    while np.linalg.norm(g(x)) > eps and k < kmax:
        B = CMI(h(x.squeeze()))
        d = -np.linalg.inv(B) @ g(x)
        alpha, _ = BLS(x, d, f, g, **BLS_params)
        x = x + alpha*d
        k += 1
        
        gradient_norm = np.round(np.linalg.norm(g(x)), precision)
        info.append([alpha, gradient_norm])
        for i in range(0, dimension):
            Xk[i].append(x[i].squeeze())
    # ================================================================================================ #
    # OUTPUT:
    print("[MN-CF] Final:",x.T)
    print("[MN-CF] Iterations:", k)
    data = pd.DataFrame(info, columns=["alpha", "||g(x)||"], dtype=np.float)
    return Xk, data

#------------------------------------------------------------------------------
# Plotter for 3D functions (only valid for two dimensions)
#------------------------------------------------------------------------------
def contour_map(f, x_axis, y_axis):
    """
    Parameters
    ----------
    f : function
    x_axis : minimization path of x
    y_axis : minimization path of y
   
    """
    xlist = np.linspace(np.min(x_axis) - 1, np.max(x_axis) + 1, 100)
    ylist = np.linspace(np.min(y_axis) - 1, np.max(y_axis) + 1, 100)

    X, Y = np.meshgrid(xlist, ylist)
    Z = f(np.array([X,Y]))
    plt.figure()
    cp = plt.contourf(X, Y, Z, cmap = 'YlGnBu', levels = 50)
    plt.colorbar(cp)
    plt.title('Descent directions')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.plot(x_axis, y_axis, c = 'red')
    plt.axis('scaled')
    plt.show()
